### ISSUE TITLE
Issues/considerations w/ running on Android on Chromebooks

### ISSUE THREAD

**Issue by fat-tire** [aaronjwood/PortAuthority#78]
On Chromebooks, Android can have its own bridged network interface with an IP not assigned by the router.  For example, on the Chromebook Pro, Android uses br0 set at 100.115.92.1 -- and when you try to portscan behind a NAT, you see this, which unless this is ipv6 seems like way too many hosts, no?

![screenshot](https://i.imgur.com/fNL9Wij.png)

It does correctly detect the WAN IP and the LAN IP is correct too.  I tried manually scanning 192.168.1.1 but though it said there 1024 hosts, it seemed to freeze when scanning.

Thoughts?

**Comment by thanset** [comment]
@fat-tire "Chromebook Pro, Android uses br0 set at 100.115.92.1", what subnet size is this?

You said you are scanning behind NAT, can you detail that a bit more the networking setup?

What is your phone's IPv4 address and in what subnet? What subnet are you trying to scan (br0?), and what size is it?  What is being NAT'd from and to here?

**Comment by aaronjwood** [comment]
This might be a bug but I need to get on implementing #59 for cases where automatic detection may not work (NAT).

**Comment by aaronjwood** [comment]
Yup, sounds like there'll need to be some extra work for the Chrome OS scenario https://developer.android.com/topic/arc/index.html#network

**Comment by fat-tire** [comment]
So as mentioned this is a Chromebook Pro using wifi behind NAT:

wlan0 is 192.168.0.223 in this case, but br0 is 100.115.92.1, which is apparently set into the Android layer of ChromeOS.  As for the subnet I'm trying scan, ideally it would be the 192.168.0.x local network-- all I've done is hit the "discover hosts" button at the bottom.

FWIW the Lan IP is listed as 192.168.0/223/0, the SSID is the correct router name. I am seeing what looks like a IPv6 WAN IP address.  Hope this answers your question...

**Comment by Bartoloni** [comment]
what is the way .. (on this cases.. or when the hosts are 65500+ ) to force scanning to... max 4096 hosts for example? (just to end the scan function in a reasonable time)

**Comment by aaronjwood** [comment]
That'd have to be an extra option built in as part of #59.

**Comment by markpitop** [comment]
I'm interested in running port authority to scan a physical network attached to the Chromebook where the wifi switched off.
When I try that I get an error "wifi not enabled" and the physical network is not scanned. 
I've had a quick look and it appears this may be coded in MainActivty:getNetworkInfo.
I'd be interested in changing this behaviour (assuming I am correct).  Could you help me with the initial steps or let me know what I'm doing wrong?

**Comment by aaronjwood** [comment]
I haven't very active on this project for a bit, my life is a little busy right now but I'm hoping I'll have some of my extra time back next year. I started [this](https://github.com/aaronjwood/PortAuthority/pull/98) to support ethernet but I don't have any devices with ethernet to test it. Once that's working and #59 is taken care of scanning networks over ethernet on Chromebooks should work. If you're interested you could build the code on the branch in my ethernet MR and try it out. I was looking for someone to test the ethernet stuff a while back.

**Comment by markpitop** [comment]
Hi Aaron,
I’d be happy to take a look, but while I know Java I’ve not done any mobile app development.
Anyway, let me spend some time looking at it see where I get. 

Thanks for coming back to me.
Mark

m.hardwick@pi-top.com  
Mobile: +447497088968  

http://pi-top.com |  London, UK

**Comment by markpitop** [comment]
Hey Arron,
I built 98 and deployed it to the Chromebook.  At first glance it’s not working.  I have the physical network connected to 192.168.0.0/24 with IP of 192.168.0.2 and wifi disabled, however Port Authority showing a lan ip of 100.115.92.2.  It is correctly showing that Wifi is disabled.  Scanning for hosts shows a dialogue saying 2147483647 hosts in subnet however scanning never starts.  It freezes at this point. 

I’m still trying to work out how to debug it on a Chromebook, and will update you once I have more specific information.

Cheers
Mark
 
http://pi-top.com |  London, UK

**Comment by aaronjwood** [comment]
Thanks for testing this! It's a huge help. Your results are good so far, if you look in the link I posted a while back in here you'll see that Chromebooks do some natting which makes it look like you're on a /1 subnet. The fact that you're able to start a scan when on Ethernet but not wireless means the logic for detecting the network connection type is working.

**Comment by aaronjwood** [comment]
```
Chromebooks run the entire Android OS in a container, similar to Docker or LXC. This means that Android will not have direct access to the system's LAN interface. Instead, IPv4 traffic will pass through an internal layer of network address translation (NAT), and IPv6 unicast traffic will be routed through an extra hop. Outbound unicast connections from an Android app to the internet should mostly work as-is; but in general, inbound connections are blocked. Multicast or broadcast packets from Android will not be forwarded to the LAN through the firewall.
```

This is why my auto subnet detection breaks on Chromebooks. I'll see if I can squeeze in some work on this over the next few weeks!

**Comment by markpitop** [comment]
Got it - that makes sense.
Don’t rush on my behalf, I’m enjoying getting to grips with the code. 
I downloaded an app that does trace route and I can see that indeed the IP address of the Chromebook is natted by an internal layer as you say. That could be tricky, is it possible to get the effective external ip from behind nat (for display in port authority)?  The Chromebook displays the ethernet port as having an address of 192.168.etc, however the app will see the natted address.

Let me know if / how I can help further.  If I find anything useful while playing with the code I’ll let you know.

M.
